http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala

```diff
@@ -24,22 +24,63 @@ import org.broadinstitute.dsde.workbench.model._
 import spray.json.{RootJsonFormat, RootJsonReader, _}
 import ca.mrvisser.sealerate
 
+import scala.util.matching.Regex
+
+sealed trait ContainerRegistry extends Product with Serializable {
+  def regex: Regex
+}
+object ContainerRegistry {
+  final case object GCR extends ContainerRegistry {
+    def regex: Regex = "(.*gcr.io)\\/(.*)\\/(.*)(\\:?.*)$".r
+
+    override def toString: String = "gcr"
+  }
+  final case object DockerHub extends ContainerRegistry {
+    def regex: Regex = ".*\\/.*$".r
+    override def toString: String = "docker hub"
+  }
+
+  val allRegistries: Set[ContainerRegistry] = sealerate.values[ContainerRegistry]
+}
+
+sealed trait JupyterDockerImage extends Product with Serializable {
+  def imageUrl: String
+  def registry: ContainerRegistry
+}
+object JupyterDockerImage {
+  final case class GCR(imageUrl: String) extends JupyterDockerImage {
+    val registry: ContainerRegistry = ContainerRegistry.GCR
+  }
+  final case class DockerHub(imageUrl: String) extends JupyterDockerImage {
+    val registry: ContainerRegistry = ContainerRegistry.DockerHub
+  }
+
+  def stringToJupyterDockerImage(imageUrl: String): Option[JupyterDockerImage] = ContainerRegistry.allRegistries.toList
+    .find(image => image.regex.pattern.asPredicate().test(imageUrl))
```

Comment:
Could the dockerhub regex match if it's a GCR image? Maybe we should ensure `allRegistries` is ordered, with more specific regexes first.

Reply:
I changed it in a way that GCR will match first. If not, then it goes to try matching dockerhub

Reply:
Nice. Perhaps add a test for that, too?